lib/datadog/tracing/span_operation.rb

```diff
@@ -424,22 +424,28 @@ def message
       # we don't want this SpanOperation to modify it further.
       def build_span
         Span.new(
-          @name && @name.dup,
+          @name.frozen? ? @name : @name.dup,
           duration: duration,
           end_time: @end_time,
           id: @id,
-          meta: meta && meta.dup,
-          metrics: metrics && metrics.dup,
+          meta: meta.dup,
+          metrics: metrics.dup,
           parent_id: @parent_id,
-          resource: @resource && @resource.dup,
-          service: @service && @service.dup,
+          resource: @resource.frozen? ? @resource : @resource.dup,
+          service: @service.frozen? ? @service : @service.dup,
           start_time: @start_time,
           status: @status,
-          type: @type && @type.dup,
+          type: @type.frozen? ? @type : @type.dup,
           trace_id: @trace_id
         )
       end
 
+      if RUBY_VERSION < '2.2' # nil.dup only fails in Ruby 2.1
+        def nil.dup
```

Comment:
Stuff like this scares me...

Reply:
My other option is to have two version of the modified methods in this PR:
```ruby
if RUBY_VERSION > '2.1'
  def initialize(
    spans,
    agent_sample_rate: nil,
    hostname: nil,
    id: nil,
    lang: nil,
    name: nil,
    origin: nil,
    process_id: nil,
    rate_limiter_rate: nil,
    resource: nil,
    root_span_id: nil,
    rule_sample_rate: nil,
    runtime_id: nil,
    sample_rate: nil,
    sampling_priority: nil,
    service: nil
  )
    @id = id
    @root_span_id = root_span_id
    @spans = spans || []
    @tags = {}

    # Set well-known tags
    self.agent_sample_rate = agent_sample_rate
    self.hostname = hostname
    self.lang = lang
    self.name = (name.frozen? ? name : name.dup)
    self.origin = (origin.frozen? ? origin : origin.dup)
    self.process_id = process_id
    self.rate_limiter_rate = rate_limiter_rate
    self.resource = (resource.frozen? ? resource : resource.dup)
    self.rule_sample_rate = rule_sample_rate
    self.runtime_id = runtime_id
    self.sample_rate = sample_rate
    self.sampling_priority = sampling_priority
    self.service = (service.frozen? ? service : service.dup)
  end
else
  def initialize(
    spans,
    agent_sample_rate: nil,
    hostname: nil,
    id: nil,
    lang: nil,
    name: nil,
    origin: nil,
    process_id: nil,
    rate_limiter_rate: nil,
    resource: nil,
    root_span_id: nil,
    rule_sample_rate: nil,
    runtime_id: nil,
    sample_rate: nil,
    sampling_priority: nil,
    service: nil
  )
    @id = id
    @root_span_id = root_span_id
    @spans = spans || []
    @tags = {}

    # Set well-known tags
    self.agent_sample_rate = agent_sample_rate
    self.hostname = hostname
    self.lang = lang
    self.name = (name.nil? || name.frozen? ? name : name.dup)
    self.origin = (name.nil? || origin.frozen? ? origin : origin.dup)
    self.process_id = process_id
    self.rate_limiter_rate = rate_limiter_rate
    self.resource = (name.nil? || resource.frozen? ? resource : resource.dup)
    self.rule_sample_rate = rule_sample_rate
    self.runtime_id = runtime_id
    self.sample_rate = sample_rate
    self.sampling_priority = sampling_priority
    self.service = (name.nil? || service.frozen? ? service : service.dup)
  end
end
```

Reply:
Leaving the nil check for Ruby > 2.1 has a 9% performance penalty:
```ruby

require 'benchmark/ips'

Benchmark.ips do |x|
  x.report("nil check") do
    resource = ''
    !resource || resource.frozen? ? resource : resource.dup
  end

  x.report("no nil check") do
    resource = ''
    resource.frozen? ? resource : resource.dup
  end

  x.compare!
end

#Comparison:
#        no nil check: 16264274.8 i/s
#           nil check: 14910035.4 i/s - 1.09x  (± 0.00) slower
```

Reply:
As long as it's justified; okay. Is this a case where "refinements" would suit?

Reply:
Genius! 💡